jax/lax/lax.py

```diff
@@ -4548,100 +4555,72 @@ def _cumred_batch_rule(prim, batched_args, batch_dims, *, axis: int):
+  ids = broadcasted_iota(onp.int32, shape, dimension)
+  outs = sort_p.bind(*(primals + (ids,)), dimension=dimension)
+  primals_out = outs[:-1]
+  sorted_ids = outs[-1]
 
-  if values_tangents is ad_util.zero:
-    values_tangents_out = ad_util.zero
-  else:
-    values_tangents_out = _sort_jvp_rule(values_tangents, keys,
-                                         dimension=dimension)
-
-  tangents_out = keys_tangents_out, values_tangents_out
-  return val_out, tangents_out
-
-def _sort_key_val_transpose_rule(t, keys, values, *, dimension):
-  t_keys, t_values = t
-  assert t_keys is ad_util.zero
-  iota = broadcasted_iota(onp.int32, keys.shape, dimension % keys.ndim)
-  _, perm = sort_key_val(keys, iota)
-  keys_result = ad_util.zero if ad.is_undefined_primal(keys) else None
-  values_result = sort_key_val(perm, t_values)[1] if ad.is_undefined_primal(values) else None
-  return [keys_result, values_result]
-
-def _sort_key_val_batch_rule(batched_args, batch_dims, *, dimension):
-  keys, values = batched_args
-  keys_bdim, values_bdim = batch_dims
-  assert keys_bdim is not None or values_bdim is not None
-  if keys_bdim == values_bdim:
-    new_dimension = dimension + (keys_bdim <= dimension)
-    return sort_key_val(keys, values, new_dimension), (keys_bdim, keys_bdim)
-  elif keys_bdim is not None and values_bdim is not None:
-    keys_trans = batching.moveaxis(keys, keys_bdim, values_bdim)
-    new_dimension = dimension + (values_bdim <= dimension)
-    return sort_key_val(keys_trans, values, new_dimension), (values_bdim, values_bdim)
-  elif keys_bdim is None:
-    broadcast_dimensions = onp.delete(onp.arange(values.ndim), values_bdim)
-    new_keys = broadcast_in_dim(keys, values.shape, broadcast_dimensions)
-    new_dimension = dimension + (values_bdim <= dimension)
-    return sort_key_val(new_keys, values, new_dimension), (values_bdim, values_bdim)
-  elif values_bdim is None:
-    broadcast_dimensions = onp.delete(onp.arange(keys.ndim), keys_bdim)
-    new_values = broadcast_in_dim(values, keys.shape, broadcast_dimensions)
-    new_dimension = dimension + (keys_bdim <= dimension)
-    return sort_key_val(keys, new_values, new_dimension), (keys_bdim, keys_bdim)
-  else:
-    assert False  # unreachable
+  dnums = GatherDimensionNumbers(
+    offset_dims=(),
+    collapsed_slice_dims=tuple(range(len(shape))),
+    start_index_map=tuple(range(len(shape))))
+  gather_id_list = []
+  for i in range(len(shape)):
+    if i == dimension:
+      gather_id_list.append(
+        broadcast_in_dim(sorted_ids, shape + (1,),
+                         broadcast_dimensions=list(range(len(shape)))))
+    else:
+      gather_id_list.append(broadcasted_iota(onp.int32, shape + (1,), dimension=i))
+  gather_ids = concatenate(gather_id_list, dimension=len(shape))
+  tangents_out = []
+  for t in tangents:
+    if t is ad_util.zero:
+      t_out = ad_util.zero
+    else:
+      t_out = gather(t, gather_ids, dnums, slice_sizes=[1] * len(shape))
```

Comment:
I just thought I'd offer this as an idea, _not_ necessarily a suggestion for you to change anything.

Instead of using a gather here, we could perhaps use NumPy indexing (to generate the gather for us), maybe something like

```python
indexer = [0] * len(shape)
indexer[dimension] = slice(None)
sorted_ids_slice = sorted_ids[indexer]

sorter = [slice(None)] * len(shape)
sorter[dimension] = sorted_ids_slice
...
  t_out = t[sorter]
```

That's probably how I'd write it but only because I'm more familiar with NumPy indexing than gather.

Reply:
Well, as far as I can tell what you wrote is in fact not correct, but the underlying idea "use NumPy indexing" is good, which I did (it was easier than I feared it was) thereby saving precious, precious vertical space.